python/cudf/cudf/core/column/column.py

```diff
@@ -934,25 +826,15 @@ def is_monotonic(self) -> bool:
 
     @property
     def is_monotonic_increasing(self) -> bool:
-        if not hasattr(self, "_is_monotonic_increasing"):
-            if self.has_nulls:
-                self._is_monotonic_increasing = False
-            else:
-                self._is_monotonic_increasing = self.as_frame()._is_sorted(
-                    ascending=None, null_position=None
-                )
-        return self._is_monotonic_increasing
+        return not self.has_nulls and self.as_frame()._is_sorted(
+            ascending=None, null_position=None
+        )
```

Comment:
Ideally this should be cached as it's definitely not free to calculate.

Reply:
In order to safely cache these properties, the cache needs to be invalidated by all operations that modify the column's values. @shwina pointed out that that is not currently happening, and I don't know how much extra work it would incur to make that safe. @shwina feel free to comment if you think there's an easy way to handle this, but it seems quite easy to mutate a column (including via methods defined in the parent classes or functions that modify the underlying data pointer, neither of which can be easily handled from within this class) and that would make caching this property unsafe.

Reply:
Yes, we are definitely not doing this. Thus, bugs like the following exist today in cudf:

```python
In [5]: s
Out[5]:
0    1
1    2
2    3
3    4
4    5
dtype: int64

In [6]: s.is_monotonic_increasing
Out[6]: True

In [7]: s[2:5] = [3, 2, 1]

In [8]: s.is_monotonic_increasing
Out[8]: True

In [9]: s
Out[9]:
0    1
1    2
2    3
3    2
4    1
dtype: int64
```

Reply:
I imagine our options are:

1. Don't cache anything. I don't mind this.
2. Cache things, but make Columns immutable. I strongly prefer this.
3. Cache things, but clear cache when the column is possibly mutated. This is bug-prone. We do something like this in `ColumnAccessor`, but that is much more localized than `Column` operations are.

Reply:
How would (2) work? Would every operation that mutates a frame necessitate copying the underlying `Column`? That seems prohibitively expensive unless copying the object amounts to modifying the underlying memory then creating a new owning Python object, but in that case you'd have to make sure that no other frames hold references to the old column to avoid any conflicts.

Reply:
@shwina can you add a github issue for that bug?

While 2) would be ideal for us as developers, there's too much code dependent on `inplace` operations. 3) sounds like the ideal case to me but is obviously very difficult / brittle / a maintenance burden.

Sounds like 1) is the best bet in the short term.

Reply:
> Would every operation that mutates a frame necessitate copying the underlying Column?

>  there's too much code dependent on inplace operations.

The vast majority of operations that modify a Frame "inplace" really perform out-of-place operations on the underlying columns. I think there's a very small set of operations that modify columns in-place; setting by slice and modifying the bitmask are really the only two I can think of.

Reply:
If I remember correct we also have a lot of situations where we swap out the buffers underneath a column which would similarly break this caching. I.E. for string "inplace" operations.

Reply:
Hmm, can you point me to an example?

Reply:
https://github.com/rapidsai/cudf/blob/d4d64c0daa3b59847276150fd4460425fcd889eb/python/cudf/cudf/_lib/column.pyx#L302-L317

This gets used underneath `_return_or_inplace` which is used underneath nearly all of the string APIs for example.

Reply:
But because this is isolated it also allows us to invalidate the cached property potentially in only a small number of places.

Reply:
I'm going to resolve this for now in favor of not caching, since determine a safe caching approach is out of scope for this PR IMO (too many possible places that would need to be changed). I've made a note on the associated issue to investigate caching further in the future.